### PR TITLE
[CINN] Upgrade generate_shape_op

### PR DESCRIPTION
**Pull request by zyfncg**
### PR types

Others

### PR changes

Others

### Description

Pcard-67164

`generate_shape_op`增加对于复杂输入符号（如`S0+1`, `Broadcast(S0, S1)`）的解析

### CODE REVIEW

File: paddle/cinn/hlir/dialect/operator/transforms/fuse_shape_ops_into_generate_shape_op_pass.cc [hunk 1]
```diff
@@ -124,17 +125,196 @@ bool MakeGenerateShapeOpAttribute(
...
+    op_set.insert(op);
+  });
+  return op_set;
+}
+
+std::vector<pir::Operation*> GetSubGraphFromOutputToInputsValue(
+    const std::vector<pir::Value>& input_values, pir::Value output_value) {
+  const std::unordered_set<pir::Operation*>& op_set =
+      GetOpSetFromOutputToInputsValue(input_values, output_value);
+  common::TopoWalker<pir::Operation*> visitor(
```
Comment (tc20042008) [inline paddle/cinn/hlir/dialect/operator/transforms/fuse_shape_ops_into_generate_shape_op_pass.cc:156]: walker

File: paddle/cinn/hlir/dialect/operator/transforms/fuse_shape_ops_into_generate_shape_op_pass.cc [hunk 2]
```diff
@@ -124,17 +125,196 @@ bool MakeGenerateShapeOpAttribute(
                                       symbol_bindings);
 }
 
+std::unordered_set<pir::Operation*> GetOpSetFromOutputToInputsValue(
+    const std::vector<pir::Value>& input_values, pir::Value output_value) {
+  std::unordered_set<pir::Operation*> op_set;
+  const std::unordered_set<pir::Value> input_value_set(input_values.begin(),
+                                                       input_values.end());
+  common::BfsWalker<pir::Operation*> walker(
+      [&](pir::Operation* node,
```
Comment (tc20042008) [inline paddle/cinn/hlir/dialect/operator/transforms/fuse_shape_ops_into_generate_shape_op_pass.cc:134]: VisitNext

File: paddle/cinn/hlir/dialect/operator/transforms/fuse_shape_ops_into_generate_shape_op_pass.cc [hunk 3]
```diff
@@ -124,17 +125,196 @@ bool MakeGenerateShapeOpAttribute(
...
+  return op_set;
+}
+
+std::vector<pir::Operation*> GetSubGraphFromOutputToInputsValue(
+    const std::vector<pir::Value>& input_values, pir::Value output_value) {
+  const std::unordered_set<pir::Operation*>& op_set =
+      GetOpSetFromOutputToInputsValue(input_values, output_value);
+  common::TopoWalker<pir::Operation*> visitor(
+      [&](pir::Operation* node,
+          const std::function<void(pir::Operation*)>& NodeHandler) {
```
Comment (tc20042008) [inline paddle/cinn/hlir/dialect/operator/transforms/fuse_shape_ops_into_generate_shape_op_pass.cc:158]: VisitUpstream

File: paddle/cinn/hlir/dialect/operator/transforms/fuse_shape_ops_into_generate_shape_op_pass.cc [hunk 4]
```diff
@@ -124,17 +125,196 @@ bool MakeGenerateShapeOpAttribute(
...
+        for (uint32_t i = 0; i < node->num_operands(); ++i) {
+          pir::Value in_value = node->operand_source(i);
+          if (in_value && in_value.defining_op()) {
+            NodeHandler(in_value.defining_op());
+          }
+        }
+      },
+      [&](pir::Operation* node,
```
Comment (tc20042008) [inline paddle/cinn/hlir/dialect/operator/transforms/fuse_shape_ops_into_generate_shape_op_pass.cc:166]: VisitDownstream

**Review by tc20042008 — APPROVED** [review]

**Review by zhangbo9674 — APPROVED** [review]

**Pull request merged**